src/java.base/share/classes/java/util/ReverseOrderSortedMapView.java

```diff
@@ -0,0 +1,464 @@
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package java.util;
+
+/**
+ * Provides a reversed-ordered view of a SortedMap. Not serializable.
+ *
+ * TODO: copy in equals and hashCode from AbstractMap
+ */
+class ReverseOrderSortedMapView<K, V> extends AbstractMap<K, V> implements SortedMap<K, V> {
+    final SortedMap<K, V> base;
+    final Comparator<? super K> cmp;
+
+    ReverseOrderSortedMapView(SortedMap<K, V> map) {
+        base = map;
+        cmp = Collections.reverseOrder(map.comparator());
+    }
+
+    public static <K, V> SortedMap<K, V> of(SortedMap<K, V> map) {
+        if (map instanceof ReverseOrderSortedMapView) {
+            return ((ReverseOrderSortedMapView<K, V>)map).base;
+        } else {
+            return new ReverseOrderSortedMapView<>(map);
+        }
+    }
+
+    // ========== Object ==========
+
+    // equals: inherited from AbstractMap
+
+    // hashCode: inherited from AbstractMap
+
+    public String toString() {
+        return toString(this, descendingEntryIterator(base));
+    }
+
+    // ========== Map ==========
+
+    public void clear() {
+        base.clear();
+    }
+
+    public boolean containsKey(Object key) {
+        return base.containsKey(key);
+    }
+
+    public boolean containsValue(Object value) {
+        return base.containsValue(value);
+    }
+
+    public V get(Object key) {
+        return base.get(key);
+    }
+
+    public boolean isEmpty() {
+        return base.isEmpty();
+    }
+
+    public V put(K key, V value) {
+        return base.put(key, value);
+    }
+
+    public void putAll(Map<? extends K, ? extends V> m) {
+        base.putAll(m);
+    }
+
+    public V remove(Object key) {
+        return base.remove(key);
+    }
+
+    public int size() {
+        return base.size();
+    }
+
+    public Set<K> keySet() {
+        return new AbstractSet<>() {
```

Comment:
At very least, please delegate `contains`, `remove`, and `clear`, to avoid O(N) complexity there. Same for `values()` and `entrySet()`

Reply:
OK. I've delegated all of these through the base map's views instead of taking shortcuts (e.g., the keyset's clear() is `base.keySet().clear()` instead of `base.clear()`) in order to preserve any special semantics the underlying map might have for these methods on its views.